src/cluster/placement/algo/mirrored.go

```diff
@@ -225,6 +231,304 @@ func (a mirroredAlgorithm) ReplaceInstances(
 	return p, nil
 }
 
+// The leaving instances were a replacement for the adding instances iff:
+//  - all shards on the leaving node are initializing and came from the adding node
+//  - all shards on the adding node are leaving (and implicitly, they should be going to the adding node--I could add a defensive check there too, come to think of it)
+//  - every leaving node has a unique adding node and every adding node has a unique leaving node (i.e. there is a bijection between them).
+// amainsd: TODO BEFORE LAND: cover the bijection case
+func instancesAreReplacement(p placement.Placement, leavingInstanceIDs []string, sourceAddingInstances []placement.Instance) bool {
+	// lookup all the instances. If one of them isn't there, we can't proceed with this case--return false
+	var leavingInstances []placement.Instance 
+	for _, instID := range leavingInstanceIDs {
+		inst, ok := p.Instance(instID)
+		if !ok {
+			return false
+		}
+		leavingInstances = append(leavingInstances, inst)
+	}
+
+	var addingInstances []placement.Instance
+	for _, addingInst := range sourceAddingInstances {
+		// N.B.: lookup here because the adding instance doesn't have shards yet.
+		inst, ok := p.Instance(addingInst.ID())
+		if !ok {
+			return false
+		}
+		addingInstances = append(addingInstances, inst)
+	}
+	
+	// every shard in leaving must have come from the same node in adding
+	validNodeSources := make(map[string]struct{})
+	for _, addingInst := range addingInstances {
+		validNodeSources[addingInst.ID()] = struct{}{}
+	}
+
+
+	for _, leavingInst := range leavingInstances {
+		var leavingNode string
+		for _, s := range leavingInst.Shards().All() {
+			switch s.State() {
+			case shard.Initializing:
+				// pass
+			default:
+				return false
+			}
+
+			// didn't come from one of the adding nodes
+			if s.SourceID() == "" {
+				return false
+			}
+
+			_, cameFromAddingNode := validNodeSources[s.SourceID()]
+			if !cameFromAddingNode {
+				return false
+			}
+
+			if leavingNode == "" {
```

Comment:
nit: why not simply initialize `leavingNode` with the `SourceID()` of the first shard? Would be less code and less logic to comprehend.